Make clickable links smarter

nibon7 opened the pull request:
# Description

Fixes #6498, when running `ls Cargo.* | find Cargo`,  `find` highlights filename with ANSI escape sequence , and the filename will be something like `"\e[37m\e[0m\e[41;37mCargo\e[0m\e[37m.lock\e[0m"`, which is not a valid path. Then we render the filename with table, and `make_clickable_link` change the filename to  `"\e]8;;\e[37m\e[0m\e[41;37mCargo\e[0m\e[37m.lock\e[0m\e\\\e[37m\e[0m\e[41;37mCargo\e[0m\e[37m.lock\e[0m\e]8;;\e\\"`, which doubles the output in the 'name' column.

[Screencast from 09-14-2022 06:15:31 PM.webm](https://user-images.githubusercontent.com/15247421/190128330-6bfbc76b-1d80-48de-80ef-86bcf73796e4.webm)


# Tests

Make sure you've done the following:

- [ ] Add tests that cover your changes, either in the command examples, the crate/tests folder, or in the /tests folder.
- [ ] Try to think about corner cases and various ways how your changes could break. Cover them with tests.
- [ ] If adding tests is not possible, please document in the PR body a minimal example with steps on how to reproduce so one can verify your change works.

Make sure you've run and fixed any issues with these commands:

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace --features=extra -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- [x] `cargo test --workspace --features=extra` to check that all the tests pass

fdncred commented:
nice work simplifying this @nibon7!!

fdncred commented:
@nibon7 I don't want to roll this back but can you work on getting ls_colors to work when finding. You can see below that the "install" files are green. However, when running find they're not green any longer.
![image](https://user-images.githubusercontent.com/343840/190151587-e4fd8d52-f2ae-43ef-ba30-3e30ff42d39f.png)

In this example the `-all.ps1` and `-all.sh`'s should be green. I think this may have to do with the ordering of the coloring where ls_colors has to happen first, then the white on red highlighting.

nibon7 commented:
> @nibon7 I don't want to roll PR this back but can you work on getting ls_colors to work when finding. 

👌

nibon7 commented:
> In this example the `-all.ps1` and `-all.sh`'s should be green. I think this may have to do with the ordering of the coloring where ls_colors has to happen first, then the white on red highlighting.

Seems `highlight_search_string` discards the lscolors.

fdncred commented:
I thought I had it working and it was just a matter of precedence. i.e. do ls_colors first, then pass that colored string to clickable links, then highlighting. Maybe not or maybe the order was different.

Maybe another option is to maybe get the color from ls_colors, highlight things with white on red, then color the rest with whatever color ls_colors returned?

nibon7 commented:
> Maybe another option is to maybe get the color from ls_colors, highlight things with white on red, then color the rest with whatever color ls_colors returned?

May we can pass the style from ls_colors, not `string_style` which is white by default. Seems it works.

https://github.com/nushell/nushell/blob/2791251268b409224496963c9b3f1726aa0beed3/crates/nu-command/src/filters/find.rs#L387-L390

nibon7 commented:
```diff
diff --git a/crates/nu-command/src/filters/find.rs b/crates/nu-command/src/filters/find.rs
index 3fe81b121..b1347f1b1 100644
--- a/crates/nu-command/src/filters/find.rs
+++ b/crates/nu-command/src/filters/find.rs
@@ -1,7 +1,7 @@
 use crate::help::highlight_search_string;
 use fancy_regex::Regex;
-use lscolors::Style as LsColors_Style;
-use nu_ansi_term::{Color::Default, Style};
+use lscolors::{Color as LsColors_Color, Style as LsColors_Style};
+use nu_ansi_term::{Color, Color::Default, Style};
 use nu_color_config::get_color_config;
 use nu_engine::{env_to_string, eval_block, CallExt};
 use nu_protocol::{
@@ -382,12 +382,16 @@ fn find_with_rest_and_highlight(
                                                 .map(LsColors_Style::to_crossterm_style)
                                                 .unwrap_or_default();
 
+                                            let new_string_style = style
+                                                .map(to_nu_style)
+                                                .unwrap_or(string_style.clone());
+
                                             let ls_colored_val =
                                                 ansi_style.apply(&val_str).to_string();
                                             let hi = match highlight_search_string(
                                                 &ls_colored_val,
                                                 &term_str,
-                                                &string_style,
+                                                &new_string_style,
                                             ) {
                                                 Ok(hi) => hi,
                                                 Err(_) => string_style
@@ -535,6 +539,51 @@ fn find_with_rest_and_highlight(
     }
 }
 
+fn to_nu_style(ls_style: &LsColors_Style) -> Style {
+    fn to_nu_color(color: &LsColors_Color) -> Color {
+        match *color {
+            LsColors_Color::Black => Color::Black,
+            LsColors_Color::Red => Color::Red,
+            LsColors_Color::Green => Color::Green,
+            LsColors_Color::Yellow => Color::Yellow,
+            LsColors_Color::Blue => Color::Blue,
+            LsColors_Color::Magenta => Color::Magenta,
+            LsColors_Color::Cyan => Color::Cyan,
+            LsColors_Color::White => Color::White,
+            LsColors_Color::BrightBlack => Color::Fixed(8),
+            LsColors_Color::BrightRed => Color::Fixed(9),
+            LsColors_Color::BrightGreen => Color::Fixed(10),
+            LsColors_Color::BrightYellow => Color::Fixed(11),
+            LsColors_Color::BrightBlue => Color::Fixed(12),
+            LsColors_Color::BrightMagenta => Color::Fixed(13),
+            LsColors_Color::BrightCyan => Color::Fixed(14),
+            LsColors_Color::BrightWhite => Color::Fixed(15),
+            LsColors_Color::Fixed(c) => Color::Fixed(c),
+            LsColors_Color::RGB(r, g, b) => Color::Rgb(r, g, b),
+        }
+    }
+
+    let mut style = Style::new();
+
+     if let Some(ref color) = ls_style.foreground {
+         style.foreground = Some(to_nu_color(color));
+     }
+
+    if let Some(ref color) = ls_style.background {
+        style.background = Some(to_nu_color(color));
+    }
+
+    style.is_bold = ls_style.font_style.bold;
+    style.is_dimmed = ls_style.font_style.dimmed;
+    style.is_italic = ls_style.font_style.italic;
+    style.is_underline = ls_style.font_style.underline;
+    style.is_reverse = ls_style.font_style.reverse;
+    style.is_hidden = ls_style.font_style.hidden;
+    style.is_strikethrough = ls_style.font_style.strikethrough;
+
+    style
+}
+
 #[cfg(test)]
 mod tests {
     use super::*;
```

nibon7 commented:
@fdncred Are you working on this, or can you test the patch above?

fdncred commented:
@nibon7 good idea. please submit a PR with that patch.
